# SQL Data Cleaning for Mobile Coverage Dataset

This **SQL** code is designed to clean and **prepare a public dataset** of mobile coverage data in **Google BigQuery**.

The primary goal is **data cleaning**, which includes creating a **new table**, **normalizing categorical features, handling missing data, and removing outliers from numerical features**.

This process is essential for **ensuring data quality and reliability** for subsequent **analysis**.

### Import libraries and modules

In [1]:
import pandas as pd
from google.cloud import bigquery

### Import function: Interactive SQL Query to Pandas DataFrame Converter

Import the custom query_df and run_query functions from the 'query_functions.py' file to execute SQL queries using a pre-configured BigQuery client object.

In [2]:
from query_functions import query_df  # Execute the query and return the output as a DataFrame
from query_functions import run_query  # Execute the query without returning a DataFrame, used for INSERT, UPDATE, DELETE, etc.

### Datasets and Tables paths to Google BigQuery

In [3]:
# Catalonian mobile coverage eu (2015-2017) --> mobile_data_2015_2017_cleaned
mobile_data_cleaned = "bq-analyst-230590.project_cat_mobile_coverage_2015_2017.mobile_data_2015_2017_cleaned"

### Creating Table and Columns

This following code creates a **copy of the 'mobile_data_2015_2017' table** by selecting specific columns and saves it as 'mobile_data_2015_2017_cleaned

In [4]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
CREATE TABLE IF NOT EXISTS `{mobile_data_cleaned}` AS
SELECT
    date,
    hour,
    lat,
    long,
    signal,
    network,
    operator,
    status,
    description,
    net,
    speed,
    satellites,
    precission,
    provider,
    activity,
    downloadSpeed,
    uploadSpeed,
    postal_code,
    town_name,
    position_geom
FROM
  `bigquery-public-data.catalonian_mobile_coverage_eu.mobile_data_2015_2017`
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Preview

In [5]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT *
FROM `{mobile_data_cleaned}` 
WHERE
    network IS NOT NULL
    AND postal_code IS NOT NULL
LIMIT 10
    """

# Execute the query
raw_data = query_df(query)

# Display data
raw_data

date      hour       lat     long  signal   network operator  status  \
0  2015-07-08  08:45:53  41.80764  1.90308      13       LCR      LCR       2   
1  2015-06-24  09:34:39  41.80858  1.89721       8       LCR      LCR       2   
2  2015-09-27  13:20:03  41.48125  2.30385      13  movistar      ONO       2   
3  2015-08-29  14:58:43  41.56906  2.39533       9  movistar      ONO       2   
4  2015-08-25  11:00:01  41.55244  2.42593      21  movistar      ONO       2   
5  2015-09-07  13:04:07  41.79837  1.95183       9  movistar      ONO       2   
6  2015-04-11  16:36:37  41.43361  2.21711      14  movistar      ONO       2   
7  2015-10-10  16:52:21  41.46863  2.24511      12  movistar      ONO       2   
8  2015-04-08  20:30:15  41.43663  2.21351      11  movistar      ONO       2   
9  2015-11-07  12:48:45  41.44589  2.22008       9  movistar      ONO       0   

            description net  speed  satellites  precission provider  \
0  STATE_EMERGENCY_ONLY  3G   19.8         0.0        14.0      gps   
1  STATE_EMERGENCY_ONLY  3G   78.6         3.0        16.0      gps   
2  STATE_EMERGENCY_ONLY  2G   18.4         4.0        41.0      gps   
3  STATE_EMERGENCY_ONLY  3G  102.2         8.0        12.0      gps   
4  STATE_EMERGENCY_ONLY  3G    2.6         7.0        37.0      gps   
5  STATE_EMERGENCY_ONLY  3G   36.4         6.0        26.0      gps   
6  STATE_EMERGENCY_ONLY  3G   12.2         4.0        16.0      gps   
7  STATE_EMERGENCY_ONLY  3G   93.9         9.0        18.0      gps   
8  STATE_EMERGENCY_ONLY  3G   15.1         5.0         8.0      gps   
9      STATE_IN_SERVICE  2G    0.6         1.0        34.0      gps   

     activity downloadSpeed uploadSpeed postal_code  town_name  \
0  IN_VEHICLE          None        None      081918    Sallent   
1  IN_VEHICLE          None        None      081918    Sallent   
2  IN_VEHICLE          None        None      080039     Alella   
3  IN_VEHICLE          None        None      080095  Argentona   
4  IN_VEHICLE          None        None      080095  Argentona   
5  IN_VEHICLE          None        None      080109      Artés   
6  IN_VEHICLE          None        None      080155   Badalona   
7  IN_VEHICLE          None        None      080155   Badalona   
8  IN_VEHICLE          None        None      080155   Badalona   
9     TILTING          None        None      080155   Badalona   

             position_geom  
0  POINT(1.90308 41.80764)  
1  POINT(1.89721 41.80858)  
2  POINT(2.30385 41.48125)  
3  POINT(2.39533 41.56906)  
4  POINT(2.42593 41.55244)  
5  POINT(1.95183 41.79837)  
6  POINT(2.21711 41.43361)  
7  POINT(2.24511 41.46863)  
8  POINT(2.21351 41.43663)  
9  POINT(2.22008 41.44589)

In [6]:
raw_data.columns

Index(['date', 'hour', 'lat', 'long', 'signal', 'network', 'operator',
       'status', 'description', 'net', 'speed', 'satellites', 'precission',
       'provider', 'activity', 'downloadSpeed', 'uploadSpeed', 'postal_code',
       'town_name', 'position_geom'],
      dtype='object')

    - Add 'province' new column

A new column, **province**, is being added to the BigQuery dataset to include province values, enhancing the dataset for regional analysis.

This new column is currently empty and will be populated based on the structure of Spanish **postal codes**, with the first two digits determining the province name. If the 'postal_code' doesn't match any of these values, it assigns 'Not defined'. Existing rows with pre-defined province values will not be affected by this update.

In [7]:
# Datasets: {mobile_data_cleaned}

# SQL query: create and add 'province' column
query = f"""
ALTER TABLE `{mobile_data_cleaned}`
ADD COLUMN province STRING;
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Populate 'province' values

In [8]:
# Datasets: {mobile_data_cleaned}

# SQL query: populate 'province' values
query = f"""
UPDATE `{mobile_data_cleaned}`
SET province = CASE
  # First two digits condition (LEFT) for postal code using CAST to maintain a consistent string data type.
  WHEN LEFT(CAST(postal_code AS STRING), 2) = '08' THEN 'Barcelona'
  WHEN LEFT(CAST(postal_code AS STRING), 2) = '25' THEN 'Lleida'
  WHEN LEFT(CAST(postal_code AS STRING), 2) = '17' THEN 'Girona'
  WHEN LEFT(CAST(postal_code AS STRING), 2) = '43' THEN 'Tarragona'
  ELSE 'Not defined'
END
WHERE province IS NULL
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Rename columns to avoid confusion and fix typos

In the original dataset, four crucial columns were present: 'net,' 'network,' 'operator,' and 'provider.' These columns provided information about the type of network (e.g., 4G, 3G, 2G), the network provider (e.g., Movistar, Orange), the specific operator (e.g., Movistar, Orange, ONO, Lowi), and the data provider (e.g., GPS, Fused, Network).

To eliminate potential confusion, particularly between the 'net' and 'network' columns, we have crafted the following SQL code to **rename** the **'network' column to 'net_provider'** and the **'provider' column to 'position_provider'**. This alteration aims to ensure the clarity and unambiguous purpose of each column and significantly enhance the overall transparency of the dataset.

In [9]:
# Datasets: {mobile_data_cleaned}

# SQL query: rename 'network' for 'net_provider'
query = f"""
ALTER TABLE `{mobile_data_cleaned}`
RENAME COLUMN network TO net_provider
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


In [10]:
# Datasets: {mobile_data_cleaned}

# SQL query: rename 'network' for 'net_provider'
query = f"""
ALTER TABLE `{mobile_data_cleaned}`
RENAME COLUMN provider TO position_provider
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


In [11]:
# Datasets: {mobile_data_cleaned}

# SQL query: fix typo in original dataset
query = f"""
ALTER TABLE `{mobile_data_cleaned}`
RENAME COLUMN precission TO precision
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Column types

In [12]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           10 non-null     object 
 1   hour           10 non-null     object 
 2   lat            10 non-null     float64
 3   long           10 non-null     float64
 4   signal         10 non-null     int64  
 5   network        10 non-null     object 
 6   operator       10 non-null     object 
 7   status         10 non-null     int64  
 8   description    10 non-null     object 
 9   net            10 non-null     object 
 10  speed          10 non-null     float64
 11  satellites     10 non-null     float64
 12  precission     10 non-null     float64
 13  provider       10 non-null     object 
 14  activity       10 non-null     object 
 15  downloadSpeed  0 non-null      object 
 16  uploadSpeed    0 non-null      object 
 17  postal_code    10 non-null     object 
 18  town_name    

In this case, '**date**' and '**hour**' have been transformed to 'object' types. However, in the original Google BigQuery dataset, their formats are 'DATE' and 'TIME', respectively. 

Just in case you need to update these formats in the original table in BigQuery, here are the required queries.

(Please note that we are displaying this data as a Python DataFrame for reference, but the actual changes must be made in the original BigQuery dataset as we will be directly querying it. We won't be working with this DataFrame for further analysis.)

In [13]:
# SQL query: in BigQuery,convert 'date' to date format and 'hour' to time format
query = f"""
    # Convert 'date' column to date data type
    UPDATE `{mobile_data_cleaned}`
    SET date = CAST(date AS DATE)
    """
# Execute the query
# run_query(query)


query = f"""
    # Convert 'date' column to date data type
    UPDATE `{mobile_data_cleaned}`
    SET hour = CAST(hour AS TIME)
    """

# Execute the query
# run_query(query)

### Categorical Features

#### Standardize data names

**Net provider**

In [14]:
# Datasets: {mobile_data_cleaned}

# SQL query: original count of unique net_provider
query = f"""
SELECT
 COUNT(DISTINCT(net_provider)) original_unique_net_provider
FROM `{mobile_data_cleaned}`
    """

# Execute the query
query_df(query)

original_unique_net_provider
0                           250

In [15]:
# Datasets: {mobile_data_cleaned}

# SQL query: example of net_provider names not standardized (Movistar)
query = f"""
SELECT
 DISTINCT(net_provider)
FROM `{mobile_data_cleaned}`
WHERE
 UPPER(net_provider) LIKE 'MO%IS%'
    """

# Execute the query
query_df(query)

net_provider
0     Movistar | Empresa
1  Movistar | Particular
2               Mobistar
3               movistar

    - Unusual characters

In [16]:
# Datasets: {mobile_data_cleaned}

# SQL query: unusual characters to take into account
query = f"""
SELECT DISTINCT net_provider
FROM `{mobile_data_cleaned}`
# Filter out net_provider values that contain non-alphanumeric characters (letters, digits, and spaces).
WHERE REGEXP_CONTAINS(net_provider, r'[^a-zA-Z0-9 ]')
    """

# Execute the query
query_df(query)

net_provider
0                      T-Mobile D
1                ANDORRA-MOBILAND
2                    TIM 98 | TIM
3              Orange F | JAZZTEL
4                     T-Mobile CZ
5                     T-Mobile.pl
6                           M-AND
7                      VodaCom-SA
8               MOBILAND | Orange
9                            AT&T
10                        O2 - UK
11                        o2 - de
12                       Si.mobil
13      vodafone ES | Profesional
14                         I.C.E.
15   Telef├│nica M├│viles Espa├▒a
16                         E-Plus
17                         TH 3G+
18                    Vodafone.de
19                     T-Mobile A
20                   STA-MOBILAND
21                  3 AT | Orange
22                NL KPN | Orange
23                    A1 | Orange
24             Movistar | Empresa
25                        O2 - CZ
26                     Digi.Mobil
27          Movistar | Particular
28                 F SFR | Orange
29          VAMOS PERU | movistar
30                    T-Mobile NL
31              Orange F | Orange
32            T-Mobile A | Orange
33                        TIM@sea
34                     VodaCom-MZ
35              Buscando serviço
36                     CHN-UNICOM
37                     Telekom.de
38                PLAY (T-Mobile)
39      France Telecom Espa├▒a SA
40  Nom├®s trucades d'emerg├¿ncia
41                    EE | Orange
42                          ?????
43                       T-Mobile
44                    VE_MOVILNET
45                  Sense servei.
46             F-Bouygues Telecom
47             MOBILAND | JAZZTEL
48                          Lowi,

This following code **standardizes the 'net_provider' names** in the table. It employs various functions like **UPPER** (to convert to uppercase), **TRIM** (to remove leading/trailing spaces), and '**=**' or **LIKE** (with **%** wildcard for pattern matching) to recognize and group similar operator names under a common name.

For example, 'movistar' and 'Mobistar' are both categorized as 'Movistar'. It ensures uniformity in the 'net_provider' column, making the data more consistent and easier to work with.

    - Standardize net_provider names

In [17]:
# Datasets: {mobile_data_cleaned}

# SQL query: Update net_provider names
query = f"""
UPDATE `{mobile_data_cleaned}`
SET net_provider = CASE
    WHEN UPPER(TRIM(net_provider)) LIKE '3%' THEN 'Three'
    WHEN UPPER(TRIM(net_provider)) LIKE '%AIRTEL%' THEN 'Airtel'
    WHEN UPPER(TRIM(net_provider)) LIKE '%B%TEL%' THEN 'Bytel'
    WHEN UPPER(TRIM(net_provider)) LIKE '%BOUYGUES%' THEN 'Bouygues Telecom'
    WHEN UPPER(TRIM(net_provider)) LIKE 'BUSCANDO %' THEN 'Sense Servei'
    WHEN UPPER(TRIM(net_provider)) = 'CABLE MOVIL' THEN 'Cable Movil'
    WHEN UPPER(TRIM(net_provider)) = 'CABLEMOVIL' THEN 'Cable Movil'
    WHEN UPPER(TRIM(net_provider)) LIKE 'CLARO%' THEN 'Claro'
    WHEN UPPER(TRIM(net_provider)) LIKE 'CUBACEL%' THEN 'Cubacel'
    WHEN UPPER(TRIM(net_provider)) LIKE 'E-%' THEN 'EE'
    WHEN UPPER(TRIM(net_provider)) LIKE '%EMER%' THEN 'Nomes Trucades Emergencies'
    WHEN UPPER(TRIM(net_provider)) LIKE 'JAZZTEL%' THEN 'Jazztel'
    WHEN UPPER(TRIM(net_provider)) LIKE 'LOWI%' THEN 'Lowi'
    WHEN LEFT(UPPER(TRIM(net_provider)), 14) = 'FRANCE TELECOM' THEN 'France Telcom Espana SA'
    WHEN UPPER(TRIM(net_provider)) LIKE 'MASMOVIL%' THEN 'Masmovil'
    WHEN UPPER(TRIM(net_provider)) LIKE '%MOBILAND%' THEN 'Mobiland'
    WHEN UPPER(TRIM(net_provider)) LIKE 'MOVILNET%' THEN 'Movilnet'
    WHEN UPPER(TRIM(net_provider)) LIKE '%MO%ISTAR%' THEN 'Movistar'
    WHEN UPPER(TRIM(net_provider)) LIKE 'MTS%' THEN 'MTS'
    WHEN UPPER(TRIM(net_provider)) LIKE '%ORANGE%' THEN 'Orange'
    WHEN UPPER(TRIM(net_provider)) = 'A1' THEN 'Orange'
    WHEN UPPER(TRIM(net_provider)) LIKE 'O2%' THEN 'O2'
    WHEN UPPER(TRIM(net_provider)) LIKE 'PE%EPHONE' THEN 'Pepephone'
    WHEN UPPER(TRIM(net_provider)) = 'PROXIMUS' THEN 'Proximus'
    WHEN UPPER(TRIM(net_provider)) LIKE 'REPUBLICA%' THEN 'Republica Movil'
    WHEN UPPER(TRIM(net_provider)) LIKE 'SENSE %' THEN 'Sense Servei'
    WHEN UPPER(TRIM(net_provider)) LIKE 'SIMYO%' THEN 'Simyo'
    WHEN UPPER(TRIM(net_provider)) LIKE 'SIN %' THEN 'Sense Servei'
    WHEN UPPER(TRIM(net_provider)) LIKE 'TIGO%' THEN 'TIGO'
    WHEN UPPER(TRIM(net_provider)) LIKE 'TIM%' THEN 'TIM'
    WHEN UPPER(TRIM(net_provider)) LIKE 'TELEF%' THEN 'Movistar'
    WHEN UPPER(TRIM(net_provider)) LIKE 'TDC%' THEN 'TDC Mobile'
    WHEN UPPER(TRIM(net_provider)) LIKE 'TELEKOM%' THEN 'Telekom'
    WHEN UPPER(TRIM(net_provider)) LIKE '%TELENOR%' THEN 'Telenor'
    WHEN UPPER(TRIM(net_provider)) LIKE '%T-MOBILE%' THEN 'T-Mobile'
    WHEN UPPER(TRIM(net_provider)) LIKE 'VIVO%' THEN 'Vivo'
    WHEN UPPER(TRIM(net_provider)) LIKE '%VODAFONE%' THEN 'Vodafone'
    WHEN UPPER(TRIM(net_provider)) = 'VF ES' THEN 'Vodafone'
    WHEN UPPER(TRIM(net_provider)) LIKE 'VODACOM%' THEN 'Vodacom'
    WHEN UPPER(TRIM(net_provider)) LIKE 'YOIGO%' THEN 'Yoigo'
    ELSE net_provider
END
WHERE net_provider IS NOT NULL;

    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Null values in net_provider

In [18]:
# Datasets: {mobile_data_cleaned}

# SQL query: Count null values and calculate the percentage
query = f"""
WITH NetProviderCount AS (
 SELECT
   COUNT(*) AS total_count_net_provider
 FROM `{mobile_data_cleaned}`
),
NullProviderCount AS (
 SELECT
   COUNT(*) AS null_count_net_provider
 FROM `{mobile_data_cleaned}`
 WHERE
   net_provider IS NULL
   OR net_provider = 'null'
)

SELECT
 null_count_net_provider,
 CONCAT(ROUND((null_count_net_provider/total_count_net_provider)*100, 2), " %") AS perc_null_net_provider
FROM NetProviderCount, NullProviderCount
"""

# Execute the query
query_df(query)

null_count_net_provider perc_null_net_provider
0                    53284                 0.45 %

     - Remove rows with specified net_provider values and NULL net_providers from the table.

In [19]:
# Datasets: {mobile_data_cleaned}

# SQL query: drop specific rows
query = f"""
DELETE FROM `{mobile_data_cleaned}`
WHERE
 net_provider IS NULL
 OR net_provider IN ('000000', '21303', '21401', '23866', '90118', '?????', 'null')
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Remove rows where the activity count for each net_provider is less than 10

In [20]:
# Datasets: {mobile_data_cleaned}

# SQL query: drop specific rows
query = f"""
DELETE FROM `{mobile_data_cleaned}`
WHERE net_provider IN (
    SELECT net_provider
    FROM `{mobile_data_cleaned}`
    GROUP BY net_provider
    HAVING COUNT(*) < 10
)
    """
# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


In [21]:
# Datasets: {mobile_data_cleaned}

# SQL query: final count of unique net_provider
query = f"""
SELECT
 COUNT(DISTINCT(net_provider)) final_unique_net_provider
FROM `{mobile_data_cleaned}`
    """

# Execute the query
query_df(query)

final_unique_net_provider
0                        122

In [22]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT
 net_provider,
 COUNT(*) activity_count
FROM `{mobile_data_cleaned}`
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20
"""

# Execute the query 
query_df(query)

net_provider  activity_count
0                     Movistar         5087640
1                       Orange         2932386
2                     Vodafone         2822286
3                        Yoigo          376842
4                     MetroPCS           95162
5                 Eroski Movil           40041
6                      Jazztel           28894
7      France Telcom Espana SA           28113
8                        TICAE           26763
9                         Lowi           21742
10                    T-Mobile           18768
11                    Mobiland           17987
12                        Vivo           17702
13                       F SFR           16402
14                       Simyo           14260
15                      GETESA           12080
16                       Bytel           11352
17                     LlamaYA            9425
18  Nomes Trucades Emergencies            8154
19                        AT&T            6822

**Operator**

In [23]:
# Datasets: {mobile_data_cleaned}

# SQL query: original count of unique net_provider
query = f"""
SELECT
 COUNT(DISTINCT(operator)) original_unique_net_operator
FROM `{mobile_data_cleaned}`
    """

# Execute the query
query_df(query)

original_unique_net_operator
0                           243

    - Standardize operator names

In [24]:
# Datasets: {mobile_data_cleaned}

# SQL query: Update operator names
query = f"""
UPDATE `{mobile_data_cleaned}`
SET operator = CASE
    WHEN UPPER(TRIM(operator)) LIKE '3%' THEN 'Three'
    WHEN UPPER(TRIM(operator)) LIKE '%AIRTEL%' THEN 'Airtel'
    WHEN UPPER(TRIM(operator)) LIKE '%B%TEL%' THEN 'Bytel'
    WHEN UPPER(TRIM(operator)) LIKE '%BOUYGUES%' THEN 'Bouygues Telecom'
    WHEN UPPER(TRIM(operator)) LIKE 'BUSCANDO %' THEN 'Sense Servei'
    WHEN UPPER(TRIM(operator)) = 'CABLE MOVIL' THEN 'Cable Movil'
    WHEN UPPER(TRIM(operator)) = 'CABLEMOVIL' THEN 'Cable Movil'
    WHEN UPPER(TRIM(operator)) LIKE 'CLARO%' THEN 'Claro'
    WHEN UPPER(TRIM(operator)) LIKE 'CUBACEL%' THEN 'Cubacel'
    WHEN UPPER(TRIM(operator)) LIKE 'E-%' THEN 'EE'
    WHEN UPPER(TRIM(operator)) LIKE '%EMER%' THEN 'Nomes Trucades Emergencies'
    WHEN UPPER(TRIM(operator)) LIKE 'JAZZTEL%' THEN 'Jazztel'
    WHEN UPPER(TRIM(operator)) LIKE 'LOWI%' THEN 'Lowi'
    WHEN LEFT(UPPER(TRIM(operator)), 14) = 'FRANCE TELECOM' THEN 'France Telcom Espana SA'
    WHEN UPPER(TRIM(operator)) LIKE 'MASMOVIL%' THEN 'Masmovil'
    WHEN UPPER(TRIM(operator)) LIKE '%MOBILAND%' THEN 'Mobiland'
    WHEN UPPER(TRIM(operator)) LIKE 'MOVILNET%' THEN 'Movilnet'
    WHEN UPPER(TRIM(operator)) LIKE '%MO%ISTAR%' THEN 'Movistar'
    WHEN UPPER(TRIM(operator)) LIKE 'MTS%' THEN 'MTS'
    WHEN UPPER(TRIM(operator)) LIKE '%ORANGE%' THEN 'Orange'
    WHEN UPPER(TRIM(operator)) = 'A1' THEN 'Orange'
    WHEN UPPER(TRIM(operator)) LIKE 'O2%' THEN 'O2'
    WHEN UPPER(TRIM(operator)) LIKE 'PE%EPHONE' THEN 'Pepephone'
    WHEN UPPER(TRIM(operator)) = 'PROXIMUS' THEN 'Proximus'
    WHEN UPPER(TRIM(operator)) LIKE 'REPUBLICA%' THEN 'Republica Movil'
    WHEN UPPER(TRIM(operator)) LIKE 'SENSE %' THEN 'Sense Servei'
    WHEN UPPER(TRIM(operator)) LIKE 'SIMYO%' THEN 'Simyo'
    WHEN UPPER(TRIM(operator)) LIKE 'SIN %' THEN 'Sense Servei'
    WHEN UPPER(TRIM(operator)) LIKE 'TIGO%' THEN 'TIGO'
    WHEN UPPER(TRIM(operator)) LIKE 'TIM%' THEN 'TIM'
    WHEN UPPER(TRIM(operator)) LIKE 'TELEF%' THEN 'Movistar'
    WHEN UPPER(TRIM(operator)) LIKE 'TDC%' THEN 'TDC Mobile'
    WHEN UPPER(TRIM(operator)) LIKE 'TELEKOM%' THEN 'Telekom'
    WHEN UPPER(TRIM(operator)) LIKE '%TELENOR%' THEN 'Telenor'
    WHEN UPPER(TRIM(operator)) LIKE '%T-MOBILE%' THEN 'T-Mobile'
    WHEN UPPER(TRIM(operator)) LIKE 'VIVO%' THEN 'Vivo'
    WHEN UPPER(TRIM(operator)) LIKE '%VODAFONE%' THEN 'Vodafone'
    WHEN UPPER(TRIM(operator)) = 'VF ES' THEN 'Vodafone'
    WHEN UPPER(TRIM(operator)) LIKE 'VODACOM%' THEN 'Vodacom'
    WHEN UPPER(TRIM(operator)) LIKE 'YOIGO%' THEN 'Yoigo'
    ELSE operator
END
WHERE operator IS NOT NULL;
    """

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - Null values in operator

In [25]:
# Datasets: {mobile_data_cleaned}

# SQL query: Count null values and calculate the percentage
query = f"""
WITH NetOperatorCount AS (
 SELECT
   COUNT(*) AS total_count_operator
 FROM `{mobile_data_cleaned}`
),
NullOperatorCount AS (
 SELECT
   COUNT(*) AS null_count_operator
 FROM `{mobile_data_cleaned}`
 WHERE
   operator IS NULL
   OR operator = 'null'
)

SELECT
 null_count_operator,
 CONCAT(ROUND((null_count_operator/total_count_operator)*100, 2), " %") AS perc_null_operator
FROM NetOperatorCount, NullOperatorCount
"""

# Execute the query
query_df(query)

null_count_operator perc_null_operator
0                    0                0 %

In [26]:
# Datasets: {mobile_data_cleaned}

# SQL query: final count of unique operator
query = f"""
SELECT
 COUNT(DISTINCT(net_provider)) final_unique_operator
FROM `{mobile_data_cleaned}`
    """

# Execute the query
query_df(query)

final_unique_operator
0                    122

In [27]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT
 operator,
 COUNT(*) activity_count
FROM `{mobile_data_cleaned}`
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20
"""

# Execute the query 
query_df(query)

operator  activity_count
0                  Movistar         4601611
1                  Vodafone         2808951
2                    Orange         2130386
3                   Jazztel          465242
4                 Pepephone          398089
5                     Yoigo          376842
6                      RACC          145208
7                     Simyo          142167
8                  MetroPCS           95162
9                    PARLEM           48442
10                      ONO           45873
11                   Tuenti           44108
12             Eroski Movil           40041
13                     Lowi           32614
14                    adamo           30888
15  France Telcom Espana SA           28113
16                    TICAE           26763
17                 T-Mobile           18768
18                 Mobiland           18007
19                     Vivo           17702

**Net**

In [28]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT
 net,
 COUNT(*) activity_record
FROM `{mobile_data_cleaned}`
GROUP BY 1
"""

# Execute the query 
query_df(query)

net  activity_record
0    3G          4045652
1    2G          2349673
2  None           763270
3    4G          4532624

    - Replace null values with 'Undefined Net'

In [29]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
UPDATE `{mobile_data_cleaned}`
SET net = CASE
  WHEN net IS NULL THEN 'Undefined net'
  ELSE net
END
WHERE net IS NULL
"""

# Execute the query 
run_query(query)

Query successfully executed, and the table has been updated.


**Position Provider**

In [30]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT
 position_provider,
 COUNT(*) activity_record
FROM `{mobile_data_cleaned}`
GROUP BY 1
"""

# Execute the query 
query_df(query)

position_provider  activity_record
0                      network              286
1                        fused          1409817
2               local_database                4
3                           19                1
4                      NETWORK               97
5             gps1503861559460                1
6                            4                2
7                          gps         10279854
8                          GPS             1125
9                         None                3
10                    disabled               25
11  2017-08-28 11:31:10.000000                1
12                          22                1
13                           7                1
14       2017-081503867352116,                1

    - Standardize provider names

In [31]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
UPDATE `{mobile_data_cleaned}`
SET position_provider = CASE
    WHEN UPPER(TRIM(position_provider)) LIKE '%GPS%' THEN 'GPS'
    WHEN UPPER(TRIM(position_provider)) LIKE '%FUSED%' THEN 'Fused'
    WHEN UPPER(TRIM(position_provider)) LIKE '%NETWORK%' THEN 'Network'
    ELSE 'Undefined'
END
WHERE 
    position_provider IS NOT NULL
    OR position_provider IS NULL
"""

# Execute the query 
run_query(query)

Query successfully executed, and the table has been updated.


In [32]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT
 position_provider,
 COUNT(*) activity_record
FROM `{mobile_data_cleaned}`
GROUP BY 1
"""

# Execute the query 
query_df(query)

position_provider  activity_record
0         Undefined               39
1             Fused          1409817
2               GPS         10280980
3           Network              383

**Postal Code** and **Town Names**

This part ensures that there are no discrepancies between postal codes and town names, and removes rows with missing values in both columns.

In [33]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT
 COUNT(*) one_valid_one_null
FROM `{mobile_data_cleaned}`
WHERE
    (postal_code IS NULL AND town_name IS NOT NULL)
    OR (postal_code IS NOT NULL AND town_name IS NULL)
"""

# Execute the query
query_df(query)

one_valid_one_null
0                   0

In [34]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT
 COUNT(*) both_features_null
FROM `{mobile_data_cleaned}`
WHERE
    postal_code IS NULL AND town_name IS NULL
"""

# Execute the query
query_df(query)

both_features_null
0              926965

    - This means that all null values in postal_code correspond to null values in `town_name. Therefore, we are going to drop all rows with null values in both features.

In [35]:
# Datasets: {mobile_data_cleaned}

# SQL query: delete specific rows
query = f"""
DELETE FROM `{mobile_data_cleaned}`
WHERE postal_code IS NULL AND town_name IS NULL
"""

# Execute the query 
run_query(query)

Query successfully executed, and the table has been updated.


**Download Speed** and **Upload Speed**

In [36]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT
 DISTINCT(downloadSpeed)
FROM `{mobile_data_cleaned}`
LIMIT 60
"""

# Execute the query
query_df(query)

downloadSpeed
0          None

In [37]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT
 DISTINCT(uploadSpeed)
FROM `{mobile_data_cleaned}`
LIMIT 60
"""

# Execute the query
query_df(query)

uploadSpeed
0        None

    - Both columns have no recorded data and only contain null values. Since they do not provide any information, we are going to drop both columns.

In [38]:
# Datasets: {mobile_data_cleaned}

# SQL query: drop specific columns
query = f"""
ALTER TABLE `{mobile_data_cleaned}`
DROP COLUMN downloadSpeed,
DROP COLUMN uploadSpeed
"""

# Execute the query 
run_query(query)

Query successfully executed, and the table has been updated.


**Description** and **Activity**

In [39]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT
 description,
 COUNT(*) activity_record
FROM `{mobile_data_cleaned}`
GROUP BY 1
LIMIT 60
"""

# Execute the query
query_df(query)

description  activity_record
0       STATE_POWER_OFF            26407
1      STATE_IN_SERVICE          1151314
2  STATE_EMERGENCY_ONLY          9586391
3  STATE_OUT_OF_SERVICE              142

In [40]:
# Datasets: {mobile_data_cleaned}

# SQL query:
query = f"""
SELECT
 activity,
 COUNT(*) activity_record
FROM `{mobile_data_cleaned}`
GROUP BY 1
LIMIT 60
"""

# Execute the query
query_df(query)

activity  activity_record
0        None               32
1     UNKNOWN           699712
2     TILTING          1092941
3       STILL          1355843
4     ON_FOOT          2065263
5  IN_VEHICLE          5349968
6  ON_BICYCLE           200495

    - Update null values to 'UNKNOWN' for the 'activity' column.

In [41]:
# Datasets: {mobile_data_cleaned}

# SQL query: update NULL values
query = f"""
UPDATE `{mobile_data_cleaned}`
SET activity = CASE 
  WHEN activity IS NULL THEN 'UNKNOWN' 
END
WHERE activity IS NULL
"""

# Execute the query 
run_query(query)

Query successfully executed, and the table has been updated.


### Numerical Features

#### Summary Statistics

In [42]:
# Datasets: {mobile_data_cleaned}

# SQL query: querying numerical features
query = f"""
SELECT
 status,
 speed,
 precision,
 signal,
 satellites
FROM `{mobile_data_cleaned}`
LIMIT 10
"""

# Execute the query
query_df(query)

status  speed  precision  signal  satellites
0       2  139.8       17.0      10         4.0
1       2   80.9       10.0      10        10.0
2       2   89.1       26.0      13         9.0
3       0   33.2       49.0      11         9.0
4       2   36.4       26.0       9         6.0
5       2    9.7        8.0      13         3.0
6       2    0.3       20.0      22        14.0
7       2   23.7       30.0      15         0.0
8       2   18.1        8.0       9         3.0
9       2   18.6        9.0      14         3.0

    - As 'status' values are directly related to the 'description' (categorical) values, we will drop the 'status' column.

In [43]:
# Datasets: {mobile_data_cleaned}

# SQL query: delete specific rows
query = f"""
ALTER TABLE `{mobile_data_cleaned}`
DROP COLUMN status
"""

# Execute the query 
run_query(query)

Query successfully executed, and the table has been updated.


Despite the option to load the entire dataset into a DataFrame and later apply the pandas .describe() function to obtain all the **summary statistics for numerical features**, we are going to manually calculate some of them **using SQL commands**.

In [44]:
# Datasets: {mobile_data_cleaned}

# SQL query: summary statistics
query = f"""
SELECT
  'speed' AS metric,
  MIN(speed) AS min_value,
  APPROX_QUANTILES(speed, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(speed, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(speed, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(speed) AS INT64) AS max_value,
  ROUND(AVG(speed), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(speed), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`

UNION ALL

SELECT
  'satellites' AS metric,
  MIN(satellites) AS min_value,
  APPROX_QUANTILES(satellites, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(satellites, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(satellites, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(satellites) AS INT64) AS max_value,
  ROUND(AVG(satellites), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(satellites), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`

UNION ALL

SELECT
  'precision' AS metric,
  MIN(precision) AS min_value,
  APPROX_QUANTILES(precision, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(precision, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(precision, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(precision) AS INT64) AS max_value,
  ROUND(AVG(precision), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(precision), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`

UNION ALL

SELECT
  'signal' AS metric,
  MIN(signal) AS min_value,
  APPROX_QUANTILES(signal, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(signal, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(signal, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(signal) AS INT64) AS max_value,
  ROUND(AVG(signal), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(signal), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`
"""

# Execute the query
query_df(query)

metric  min_value  percentile_25  median  percentile_75  \
0  satellites        0.0            2.0     4.0            7.0   
1   precision        0.0           10.0    17.0           30.0   
2      signal        0.0            8.0    12.0           18.0   
3       speed        0.0            1.4     7.2           39.3   

         max_value    avg_value    std_value  
0   11503299477926   1068783.46   3506348267  
1  201503299477926  28152604.87  68773279817  
2               99        13.20            7  
3              255        25.85           35

- **APPROX_QUANTILES(column, percentil_divisions)[OFFSET(percentil_index)]**:

Is a SQL construct used to estimate quantiles for the specified column, with the option to retrieve a specific quantile based on its position within the array of calculated quantiles. This is particularly useful for analyzing data distributions and summarizing data into percentiles or quartiles.

    - Handling 'precision' and 'satellites' possible outliers 

In [45]:
# Datasets: {mobile_data_cleaned}

# SQL query: precision values ordered DESC to check maximum values
query = f"""
SELECT
 CAST(precision AS INT64) precision
FROM `{mobile_data_cleaned}`
ORDER BY 1 DESC
LIMIT 10
"""

# Execute the query
query_df(query)

precision
0  201503299477926
1  101503867043000
2             5304
3             3799
4             3400
5             3400
6             3400
7             3400
8             3400
9             3400

In [46]:
# Datasets: {mobile_data_cleaned}

# SQL query: satellites values ordered DESC to check maximum values
query = f"""
SELECT
 CAST(satellites AS INT64) satellites
FROM `{mobile_data_cleaned}`
ORDER BY 1 DESC
LIMIT 10
"""

# Execute the query
query_df(query)

satellites
0  11503299477926
1              42
2              29
3              29
4              29
5              29
6              29
7              29
8              28
9              28

    - Handling outliers in 'precision' colum

Now we are going to calculate various percentiles (25th, 50th, 75th, 90th, 95th, 99th) for the 'precision' column in the table and display their values.

In [47]:
# Datasets: {mobile_data_cleaned}

# SQL query: satellites values ordered DESC to check maximum values
query = f"""
WITH Percentiles AS (
  SELECT
    APPROX_QUANTILES(precision, 100)[OFFSET(25)] AS percentile_25,
    APPROX_QUANTILES(precision, 100)[OFFSET(50)] AS median,
    APPROX_QUANTILES(precision, 100)[OFFSET(75)] AS percentile_75,
    APPROX_QUANTILES(precision, 100)[OFFSET(90)] AS percentile_90,
    APPROX_QUANTILES(precision, 100)[OFFSET(95)] AS percentile_95,
    APPROX_QUANTILES(precision, 100)[OFFSET(99)] AS percentile_99
  FROM `{mobile_data_cleaned}`
)
SELECT percentile_25, median, percentile_75, percentile_90, percentile_95, percentile_99
FROM Percentiles;
"""

# Execute the query
query_df(query)

percentile_25  median  percentile_75  percentile_90  percentile_95  \
0           10.0    17.0           30.0           48.0           62.0   

   percentile_99  
0          119.0

In [49]:
# Datasets: {mobile_data_cleaned}

# SQL query: precision equal or greater or greater than 119
query = f"""
SELECT
 COUNT(*) precision_eq_or_gr_t_119
FROM `{mobile_data_cleaned}`
WHERE precision >= 119
"""

# Execute the query
query_df(query)

precision_eq_or_gr_t_119
0                    109904

    - Percentile 99: the top 1% of values (109904 rows) have precision values equal or greater than 119.

    - How would the removal of these outliers affect the data distribution, and would this impact be consistent across different provinces?

This following query calculates and compares outlier percentages in different provinces based on the 'precision' column, where values equal or above 119 are considered outliers. It provides insights into how outliers are distributed among provinces.

In [50]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT
  province,
  COUNT(*) AS province_records,
  SUM(CASE WHEN precision >= 119 THEN 1 ELSE 0 END) AS province_outliers,
  ROUND((SUM(CASE WHEN precision >= 119 THEN 1 ELSE 0 END) / COUNT(*)) * 100, 2) AS percentage_outliers
FROM `{mobile_data_cleaned}`
GROUP BY 1
ORDER BY 1
"""

# Execute the query
query_df(query)

province  province_records  province_outliers  percentage_outliers
0  Barcelona           7367393              79546                 1.08
1     Girona           1391076              14982                 1.08
2     Lleida           1003705               4575                 0.46
3  Tarragona           1002080              10801                 1.08

    - Barcelona, Girona and Tarragona would have the same impact (1.08% of their data) if these outliers are removed. Only Lleida would have a smaller effect (0.46%).
    
    - We are going to proceed with the deletion of 'precision' values equal or greater than 119.

In [51]:
# Datasets: {mobile_data_cleaned}

# SQL query: delete outliers in 'precision' column
query = f"""
DELETE FROM `{mobile_data_cleaned}`
WHERE precision >= 119
"""

# Execute the query
run_query(query)

Query successfully executed, and the table has been updated.


    - As a result of removing the 'precision' outliers, the outlier in the 'satellites' column, which had a value of 11503299477926, has also been eliminated from the dataset.

In [52]:
# Datasets: {mobile_data_cleaned}

# SQL query: satellites values ordered DESC to check maximum values
query = f"""
SELECT
 CAST(satellites AS INT64) satellites
FROM `{mobile_data_cleaned}`
ORDER BY 1 DESC
LIMIT 10
"""

# Execute the query
query_df(query)

satellites
0          42
1          29
2          29
3          29
4          29
5          29
6          29
7          28
8          28
9          28

    - Summary Statistics

In [53]:
# Datasets: {mobile_data_cleaned}

# SQL query: summary statistics
query = f"""
SELECT
  'speed' AS metric,
  MIN(speed) AS min_value,
  APPROX_QUANTILES(speed, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(speed, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(speed, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(speed) AS INT64) AS max_value,
  ROUND(AVG(speed), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(speed), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`

UNION ALL

SELECT
  'satellites' AS metric,
  MIN(satellites) AS min_value,
  APPROX_QUANTILES(satellites, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(satellites, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(satellites, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(satellites) AS INT64) AS max_value,
  ROUND(AVG(satellites), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(satellites), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`

UNION ALL

SELECT
  'precision' AS metric,
  MIN(precision) AS min_value,
  APPROX_QUANTILES(precision, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(precision, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(precision, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(precision) AS INT64) AS max_value,
  ROUND(AVG(precision), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(precision), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`

UNION ALL

SELECT
  'signal' AS metric,
  MIN(signal) AS min_value,
  APPROX_QUANTILES(signal, 100)[OFFSET(25)] AS percentile_25,
  APPROX_QUANTILES(signal, 100)[OFFSET(50)] AS median,
  APPROX_QUANTILES(signal, 100)[OFFSET(75)] AS percentile_75,
  CAST(MAX(signal) AS INT64) AS max_value,
  ROUND(AVG(signal), 2) AS avg_value,
  CAST(ROUND(STDDEV_POP(signal), 2) AS INT64) AS std_value
FROM `{mobile_data_cleaned}`
"""

# Execute the query
query_df(query)

metric  min_value  percentile_25  median  percentile_75  max_value  \
0      signal        0.0            8.0    12.0           18.0         99   
1  satellites        0.0            2.0     5.0            7.0         42   
2       speed        0.0            1.4     7.3           39.7        255   
3   precision        0.0           10.0    17.0           29.0        118   

   avg_value  std_value  
0      13.22          7  
1       4.97          4  
2      25.97         35  
3      21.91         18

### Save Cleaned Table

If required, the fully cleaned dataset could be stored under the name 'mobile_data_2015_2017_cleaned_final' for use in subsequent analysis and reporting.

In [54]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
CREATE OR REPLACE TABLE `bq-analyst-230590.project_cat_mobile_coverage_2015_2017.mobile_data_2015_2017_cleaned_final`
AS
SELECT *
FROM `{mobile_data_cleaned}`
"""

# Execute the query
# run_query(query)

### Cleaned Table

In [55]:
# Datasets: {mobile_data_cleaned}

# SQL query: 
query = f"""
SELECT *
FROM `{mobile_data_cleaned}`
#LIMIT 20
"""

# Execute the query
cleaned_data = query_df(query)

# Display data
display(cleaned_data.head(), cleaned_data.tail(), cleaned_data.shape, cleaned_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10654350 entries, 0 to 10654349
Data columns (total 18 columns):
 #   Column             Dtype  
---  ------             -----  
 0   date               object 
 1   hour               object 
 2   lat                float64
 3   long               float64
 4   signal             int64  
 5   net_provider       object 
 6   operator           object 
 7   description        object 
 8   net                object 
 9   speed              float64
 10  satellites         float64
 11  precision          float64
 12  position_provider  object 
 13  activity           object 
 14  postal_code        object 
 15  town_name          object 
 16  position_geom      object 
 17  province           object 
dtypes: float64(5), int64(1), object(12)
memory usage: 1.4+ GB


date      hour       lat     long  signal net_provider  operator  \
0  2016-06-08  12:12:35  42.42245  2.93821       4     Vodafone  Vodafone   
1  2015-06-16  14:51:10  42.02592  2.87216      15     Vodafone  Vodafone   
2  2016-07-23  18:12:13  42.22272  2.74340       9     Vodafone  Vodafone   
3  2016-09-05  12:02:11  42.19797  2.19390       4     Vodafone  Vodafone   
4  2015-05-19  16:49:58  41.89019  2.78493      13     Vodafone  Vodafone   

            description net  speed  satellites  precision position_provider  \
0  STATE_EMERGENCY_ONLY  3G    2.3         7.0       11.0               GPS   
1      STATE_IN_SERVICE  3G    0.1         2.0       19.0               GPS   
2  STATE_EMERGENCY_ONLY  3G   31.0        12.0        6.0               GPS   
3  STATE_EMERGENCY_ONLY  3G    0.0        11.0       11.0               GPS   
4  STATE_EMERGENCY_ONLY  3G   69.2         6.0       21.0               GPS   

     activity postal_code               town_name            position_geom  \
0     UNKNOWN      170412              Cantallops  POINT(2.93821 42.42245)   
1     UNKNOWN      170499                   Celrà  POINT(2.87216 42.02592)   
2       STILL      170985         Maià de Montcal   POINT(2.7434 42.22272)   
3     ON_FOOT      171479                  Ripoll   POINT(2.1939 42.19797)   
4  IN_VEHICLE      171501  Riudellots de la Selva  POINT(2.78493 41.89019)   

  province  
0   Girona  
1   Girona  
2   Girona  
3   Girona  
4   Girona

date      hour       lat     long  signal net_provider  \
10654345  2015-11-03  05:58:44  41.58343  0.88596      14     Vodafone   
10654346  2016-09-06  15:15:04  41.16102  1.10252      12     Vodafone   
10654347  2016-01-31  14:35:56  42.25248  1.86068       8     Movistar   
10654348  2015-11-26  01:41:00  41.39091  2.16298       2     Movistar   
10654349  2015-10-28  04:59:54  41.41830  2.20058      10     Movistar   

           operator           description net  speed  satellites  precision  \
10654345   Vodafone  STATE_EMERGENCY_ONLY  3G    1.2         8.0       32.0   
10654346   Vodafone  STATE_EMERGENCY_ONLY  4G   12.8         0.0       50.0   
10654347  Pepephone      STATE_IN_SERVICE  2G    0.9         6.0       87.0   
10654348  Pepephone  STATE_EMERGENCY_ONLY  4G   16.3         4.0       48.0   
10654349  Pepephone  STATE_EMERGENCY_ONLY  4G   24.2         3.0       48.0   

         position_provider    activity postal_code  town_name  \
10654345               GPS     UNKNOWN      251351  Miralcamp   
10654346             Fused  IN_VEHICLE      431233       Reus   
10654347               GPS  IN_VEHICLE      080168       Bagà   
10654348               GPS  IN_VEHICLE      080193  Barcelona   
10654349               GPS  IN_VEHICLE      080193  Barcelona   

                    position_geom   province  
10654345  POINT(0.88596 41.58343)     Lleida  
10654346  POINT(1.10252 41.16102)  Tarragona  
10654347  POINT(1.86068 42.25248)  Barcelona  
10654348  POINT(2.16298 41.39091)  Barcelona  
10654349   POINT(2.20058 41.4183)  Barcelona

(10654350, 18)

None